### ISSUE TITLE
License server issues

### ISSUE THREAD

**Issue by JOJ0** [jorritfolmer/puppet-splunk#53]
Hi,
first things first: Thanks for this very useful puppet module! :-)
I am having a very hard time figuring out how to get my license server and slaves to trust each other. This is what I see in splunkd.log on my license master (which is also the indexer cluster master):

```
10-30-2020 19:32:38.391 +0100 ERROR LMMasterRestHandler - path=/masterlm/usage: Signature mismatch between license slave=172.27.42.1 and this License Master. Please make sure that the pass4SymmKey setting in server.conf, under [general], is the same for the License Master and all its slaves from ip=172.27.42.1
```

My configuration looks similar to example 4 in README.md.

This is the configuration all my splunk machines get:
```
splunk::service:
    ensure:  'running'
    enable:  'true'
splunk::admin:
    hash:  '$6$tRxxxxxxsecret'
    fn:    'Splunk Admin'
    email: 'splunk-admin@my-domain.com'
splunk::lm: 'cluster-and-license-master-fqdn:8089'
```

This is what my cluster master gets:
```
splunk::httpport:         8000
splunk::tcpout:           'indexer_discovery'
splunk::clustering:
    mode:                     'master'
    indexer_discovery:        'true'
    replication_factor:       2
    search_factor:            2
    site_replication_factor:  'origin:1,total:2'
    site_search_factor:       'origin:1,total:2'
    thissite:                 'site1'
    available_sites:          'site1,site2'
    pass4symmkey:             'plaintextsecret'
```

This is what my cluster indexers get:

```
splunk::clustering:
    thissite: 'site1'
    forwarder_site_failover: 'site1:site2'
```

```
splunk::inputport:        9997
splunk::httpport:         8000
splunk::replication_port: 8080
splunk::clustering:
    mode: 'slave'
    cm: 'cluster-and-license-master-fqdn:8089'
    pass4symmkey: 'plaintextsecret'
```

I set merge settings for splunk::clustering like this because indexer configuration is kept in two different hiera levels:
```
    splunk::clustering:
        merge: deep
```

I am running out of ideas how to debug this further. I can see that from time to time on puppetruns that pass4SymmKey is set back to changeme or plaintextsecret. I am afraid I don't understand closely enough when plaintextsecret is exchanged with a real hashed key. Does splunk do that or does the puppet module do that?

Maybe you could point me in a direction?
Thanks again for this module! Configuring my splunk setup without it would be impossible I guess! :-)

all the best
Jojo

**Comment by jorritfolmer** [comment]
To be honest I have no additional clue that may help you. Thanks for the kind words by the way!
Maybe @sickbock can offer some insights? He seems to be using hiera merging more often than I do.

**Comment by sickbock** [comment]
My guess is you need to fork the project, specify the desired merge type in data/common.yaml, and create a pull request. I'm only using splunk::admin and splunk::auth with multiple yaml files because I needed the passwords in eyaml files. 

About the pass4symmkey and other sensitive data (except the admin password): Splunk encrypts it with its "secret", make sure you use the same secret on all your servers.

**Comment by JOJ0** [comment]
Hi @jorritfolmer and @sickbock,
thanks a lot for your replies. I could solve it by now. Indeed it had to do with hiera merging or at least with "who gets what settings from different hiera levels".

Well, actually I am not a 100% sure but let's write this down here anyway: To work around the problem I first commented out the splunk::lm:... line, then set up my license master and all the license clients manually, while doing it I realized that machine 'cluster-and-license-master-fqdn' also was set up to use a remote license master (it's own fqdn) because I had set the following things for _all_ my splunk nodes, including the machine supposed to run the license master and read its license locally:

> This is the configuration all my splunk machines get:
> 
> ```
> splunk::service:
>     ensure:  'running'
>     enable:  'true'
> splunk::admin:
>     hash:  '$6$tRxxxxxxsecret'
>     fn:    'Splunk Admin'
>     email: 'splunk-admin@my-domain.com'
> splunk::lm: 'cluster-and-license-master-fqdn:8089'

I now put the splunk::lm line to all my splunk nodes but not on the license master itself and it still works after puppet runs.
